# Reading Files with Pandas

In this section we will use Pandas to explore tabular datasets (mainly in CSV form)

## Input Data

Our input data contains various data files from the New York Times about Covid-19 spread across the entire world. You can download the files from their GitHub repository https://github.com/nytimes/covid-19-data

As a starter we will look into the file facilities.csv which contain the details of Covid19 spread in correctional facilities. During the pandemic outbreaks were common in confined facilities such as prisons, carehomes, and nursing homes. 

## Reading a CSV File In Pandas

For reading a CSV file we utilize the read_csv() function from Pandas. The read_csv function expects a file-path and it also has many optional parameters to modify the way in which the file is read. 

In [1]:
import pandas as pd

If you want to know the details about the function read_csv you can do the following

In [2]:
pd.read_csv?

Signature:
pd.read_csv(
    filepath_or_buffer: 'FilePath | ReadCsvBuffer[bytes] | ReadCsvBuffer[str]',
    sep=<no_default>,
    delimiter=None,
    header='infer',
    names=<no_default>,
    index_col=None,
    usecols=None,
    squeeze=None,
    prefix=<no_default>,
    mangle_dupe_cols=True,
    dtype: 'DtypeArg | None' = None,
    engine: 'CSVEngine | None' = None,
    converters=None,
    true_values=None,
    false_values=None,
    skipinitialspace=False,
    skiprows=None,
    skipfooter=0,
    nrows=None,
    na_values=None,
    keep_default_na=True,
    na_filter=True,
    verbose=False,
    skip_blank_lines=True,
    parse_dates=None,
    infer_datetime_format=False,
    keep_date_col=False,
    date_parser=None,
    dayfirst=False,
    cache_dates=True,
    iterator=False,
    chunksize=None,
    compression: 'CompressionOptions' = 'infer',
    thousands=None,
    decimal: 'str' = '.',
    lineterminator=None,
    quotechar='"',
    quoting=0,
    doublequote=True,
    esc

In [3]:
prisonData = pd.read_csv('data/facilities.csv')  #We are passing just the filepath here

Quick question. What is the datatype for prisonData. 

If you want to just see the first few records or last few records you can use the methods head and tail respectively

In [4]:
prisonData.head()

nyt_id                                      facility_name  \
0  F3EFE858                      Alex City Work Release prison   
1  5B910220      Alabama Therapeutic Education Facility prison   
2  02FB1675                         Bibb Correctional Facility   
3  6378F6C4  Birmingham Women's Community Based Facility an...   
4  EAABF900                      Bullock Correctional Facility   

                 facility_type facility_city facility_county  \
0    Low-security work release     Alex City           Coosa   
1  State rehabilitation center    Columbiana          Shelby   
2                 State prison         Brent            Bibb   
3                 State prison    Birmingham       Jefferson   
4                 State prison      Bessemer         Bullock   

   facility_county_fips facility_state  facility_lng  facility_lat  \
0                  1037        Alabama    -86.009015     32.904507   
1                  1117        Alabama    -86.624067     33.180755   
2                  1007        Alabama    -87.162781     32.920754   
3                  1073        Alabama    -86.808344     33.531101   
4                  1011        Alabama    -85.673927     32.147144   

   latest_inmate_population  max_inmate_population_2020  total_inmate_cases  \
0                     188.0                         NaN                  77   
1                     272.0                         NaN                  11   
2                    1725.0                      1825.0                 164   
3                     192.0                         NaN                  17   
4                    1477.0                      1577.0                 162   

   total_inmate_deaths  total_officer_cases  total_officer_deaths note  
0                    0                   17                   0.0  NaN  
1                    1                    2                   0.0  NaN  
2                    3                   61                   0.0  NaN  
3                    0                   28                   0.0  NaN  
4                    5                   80                   1.0  NaN

In [5]:
prisonData.tail()

nyt_id                    facility_name   facility_type facility_city  \
2634  15289545        North Lake federal prison  Federal prison       Baldwin   
2635  1558C2BF            Rivers federal prison  Federal prison        Winton   
2636  C9CF62B9     Reeves County federal prison  Federal prison         Pecos   
2637  364869B9        Flightline federal prison  Federal prison    Big Spring   
2638  5CFB7978  Moshannon Valley federal prison  Federal prison   Philipsburg   

     facility_county  facility_county_fips  facility_state  facility_lng  \
2634            Lake                 26085        Michigan    -85.839287   
2635        Hertford                 37091  North Carolina    -76.958751   
2636          Reeves                 48389           Texas   -103.493817   
2637          Howard                 48227           Texas   -101.521236   
2638      Clearfield                 42033    Pennsylvania    -78.242177   

      facility_lat  latest_inmate_population  max_inmate_population_2020  \
2634     43.928551                    1614.0                      1614.0   
2635     36.403668                       6.0                      1255.0   
2636     31.423563                    1005.0                         NaN   
2637     32.224310                    1652.0                         NaN   
2638     40.921117                      94.0                      1774.0   

      total_inmate_cases  total_inmate_deaths  total_officer_cases  \
2634                 125                    2                    0   
2635                  68                    1                    0   
2636                  46                    1                    0   
2637                  34                    1                    0   
2638                 197                    4                    0   

      total_officer_deaths note  
2634                   0.0  NaN  
2635                   0.0  NaN  
2636                   0.0  NaN  
2637                   0.0  NaN  
2638                   0.0  NaN

How many rows and columns are there in our prison dataset

Can you enumerate the columns and the index in this dataset

What are the datatypes for each columns

Can you select the columns 'total_inmate_cases' from the dataset. What is the datatype for the column?

Can you select the columns 'nyt_id', 'facility_name','facility_state','latest_inmate_population' and 'total_inmate_cases' from the dataset

Some basic descriptive stats for the dataset

In [6]:
prisonData.describe()

facility_county_fips  facility_lng  facility_lat  \
count           2639.000000   2639.000000   2639.000000   
mean           29186.422887    -91.193161     37.280626   
std            15809.439620     14.836097      5.437975   
min             1003.000000   -165.412569     18.423135   
25%            13233.000000    -97.649683     33.422542   
50%            30017.000000    -86.812013     37.606861   
75%            42033.000000    -81.115561     40.933270   
max            72061.000000    -66.112209     64.833337   

       latest_inmate_population  max_inmate_population_2020  \
count               1593.000000                  838.000000   
mean                 822.386692                 1269.529833   
std                  862.009973                  936.057182   
min                    1.000000                   27.000000   
25%                  163.000000                  602.500000   
50%                  579.000000                 1066.500000   
75%                 1197.000000                 1681.000000   
max                 8616.000000                 6150.000000   

       total_inmate_cases  total_inmate_deaths  total_officer_cases  \
count         2639.000000          2639.000000          2639.000000   
mean           199.431603             0.983327            42.561955   
std            420.344387             2.982973            89.408213   
min              0.000000             0.000000             0.000000   
25%              8.000000             0.000000             0.000000   
50%             59.000000             0.000000             9.000000   
75%            226.500000             0.000000            50.500000   
max          12290.000000            45.000000          1718.000000   

       total_officer_deaths  
count           2638.000000  
mean               0.083776  
std                0.363528  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max                6.000000

While running descriptive stats on some of the columns such as facility_county_fips doesn't make any sense, it can easily provide an overview for columns such as total_inmate_cases (mean = 199.431603,max = 12290.000000). 

Can you just print the descriptive statistics for total_inmate_cases and total_inmate_deaths

## Some basic calculations

Create a new field cases_per_population which is calculated as total_inmate_cases/latest_inmate_population

Create a new field deaths_per_population which is calculated as total_inmate_deaths/latest_inmate_population

Find the correctional facility with the maximum number of inmate cases?

Find the correctional facility with the maximum number of inmate deaths?

## Selections

 Can you select the first 10 rows of this dataset. 

Can you select the last 100 rows for the total_inmate_cases.

Can you select the last 100 rows for the total_inmate_cases and total_inmate_deaths.

Can you select the last 100 rows for the total_inmate_cases and calculate the mean.

## Filtering and updating data

Let us find all facilities with more than 1000 inmate cases

In [7]:
prisonData[prisonData.total_inmate_cases>1000]

nyt_id                           facility_name   facility_type  \
34    7120943C         Goose Creek Correctional Center    State prison   
50    AA807759  Arizona State Prison Complex – Douglas    State prison   
51    91B22ADD                    Eyman Complex prison    State prison   
52    0CF96DB1     Arizona State Prison Complex – Yuma    State prison   
58    E830C278    Arizona State Prison Complex – Lewis    State prison   
...        ...                                     ...             ...   
2568  233CC887          Pollock federal prison complex  Federal prison   
2588  79BEB932               Seagoville federal prison  Federal prison   
2599  4A388A9A      Terre Haute federal prison complex  Federal prison   
2608  D9A37098           Tucson federal prison complex  Federal prison   
2609  20C2E350      Victorville federal prison complex  Federal prison   

     facility_city    facility_county  facility_county_fips facility_state  \
34         Wasilla  Matanuska-Susitna                  2170         Alaska   
50         Douglas            Cochise                  4003        Arizona   
51        Florence              Pinal                  4021        Arizona   
52        San Luis               Yuma                  4027        Arizona   
58         Buckeye           Maricopa                  4013        Arizona   
...            ...                ...                   ...            ...   
2568       Pollock              Grant                 22043      Louisiana   
2588    Seagoville             Dallas                 48113          Texas   
2599   Terre Haute               Vigo                 18167        Indiana   
2608        Tucson               Pima                  4019        Arizona   
2609   Victorville     San Bernardino                  6071     California   

      facility_lng  facility_lat  latest_inmate_population  \
34     -149.990022     61.360095                    1293.0   
50     -109.600147     31.445500                    1848.0   
51     -111.337569     33.031474                    5362.0   
52     -114.643082     32.494131                    4316.0   
58     -112.644365     33.201281                    4617.0   
...            ...           ...                       ...   
2568    -92.432853     31.465159                    2590.0   
2588    -96.567986     32.655933                    1798.0   
2599    -87.455019     39.410206                    2291.0   
2608   -110.865276     32.084730                    1586.0   
2609   -117.364873     34.567075                    3310.0   

      max_inmate_population_2020  total_inmate_cases  total_inmate_deaths  \
34                        1348.0                1041                    0   
50                        2132.0                1163                    0   
51                        5471.0                2023                    5   
52                        4862.0                2010                    6   
58                        4617.0                1310                    2   
...                          ...                 ...                  ...   
2568                      2754.0                1211                    0   
2588                      2004.0                1336                    5   
2599                      2624.0                1248                    6   
2608                      1931.0                1067                   10   
2609                      4210.0                1696                    3   

      total_officer_cases  total_officer_deaths note  
34                      0                   0.0  NaN  
50                      0                   0.0  NaN  
51                      0                   0.0  NaN  
52                      0                   0.0  NaN  
58                      0                   0.0  NaN  
...                   ...                   ...  ...  
2568                    0                   0.0  NaN  
2588                    0                   0.0  NaN  
2599             

Now can you try to create a subset of our dataset containing only facilities from Ohio?

Can you try to find out facilties in Franklin County, Ohio with more than 30 inmate cases. 

Which is the facility in Ohio with most number of inamte deaths

## Dealing with missing data

Data gaps are common in real-world datasets. Let us see the records with missing latest_inmate_population.

In [8]:
prisonData[prisonData['latest_inmate_population'].isna()]

nyt_id                                      facility_name  \
39    ED29F8B4                         Palmer Correctional Center   
57    07E5D203                            Maricopa Reentry Center   
61    26CFE1E6                                Pima Reentry Center   
79    9AA13461             Northwest Arkansas Work Release prison   
88    7C68181A  Arkansas Juvenile Assessment and Treatment Center   
...        ...                                                ...   
2620  786CB2A3             Roseville federal prison halfway house   
2622  175CAB9F       Watkinson House federal prison halfway house   
2624  87DBAF80         Parsons House federal prison halfway house   
2625  E3A87C51             Flagstaff federal prison halfway house   
2626  A7B2A8A7          Garden Grove federal prison halfway house   

                  facility_type facility_city    facility_county  \
39                 State prison        Palmer  Matanuska-Susitna   
57                 State prison       Phoenix           Maricopa   
61                 State prison        Tucson               Pima   
79    Low-security work release    Springdale         Washington   
88     State juvenile detention     Alexander             Saline   
...                         ...           ...                ...   
2620      Federal halfway house     Roseville             Ramsey   
2622      Federal halfway house      Hartford           Hartford   
2624      Federal halfway house     Milwaukee          Milwaukee   
2625      Federal halfway house     Flagstaff           Coconino   
2626      Federal halfway house  Garden Grove             Orange   

      facility_county_fips facility_state  facility_lng  facility_lat  \
39                    2170         Alaska   -149.007972     61.691417   
57                    4013        Arizona   -112.120433     33.707162   
61                    4019        Arizona   -110.990714     32.204199   
79                    5143       Arkansas    -94.135712     36.174948   
88                    5125       Arkansas    -92.494695     34.637731   
...                    ...            ...           ...           ...   
2620                 27123      Minnesota    -93.124231     44.994233   
2622                  9003    Connecticut    -72.689572     41.772700   
2624                 55079      Wisconsin    -87.960541     43.039124   
2625                  4005        Arizona   -111.640458     35.187256   
2626                  6059     California   -117.936021     33.775822   

      latest_inmate_population  max_inmate_population_2020  \
39                         NaN                         NaN   
57                         NaN                         NaN   
61                         NaN                         NaN   
79                         NaN                         NaN   
88                         NaN                         NaN   
...                        ...                         ...   
2620                       NaN                         NaN   
2622                       NaN                         NaN   
2624                       NaN                         NaN   
2625                       NaN                         NaN   
2626                       NaN                         NaN   

      total_inmate_cases  total_inmate_deaths  total_officer_cases  \
39                     0                    0                    0   
57                     0                    0                    0   
61                     0                    0                    0   
79                    69                    0                    4   
88                    63                    0                   67   
...                  ...                  ...                  ...   
2620                   8                    0                    0   
2622                   1                    0                    0   
2624                   1                    0                    0   
2625                   1                

Wowwww that's a lot of records. 

Now can you filter out the records that doesnot have a null value for latest_inmate_population

## Data types

Let us see the datatype for the column facility_county_fips

In [9]:
prisonData['facility_county_fips'].dtype

dtype('int64')

Can you convert this column to string?

## Finding Unique Values

We can use unique() method to extract out unique values from a column. For example let's extract out the unique values from facility_state column. 

In [10]:
prisonData['facility_state'].unique()

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'Florida', 'Georgia',
       'Hawaii', 'Idaho', 'Texas', 'Illinois', 'Indiana', 'Iowa',
       'Kansas', 'Kentucky', 'Louisiana', 'Maine', 'Maryland',
       'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi',
       'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire',
       'New Jersey', 'New Mexico', 'New York', 'North Carolina',
       'North Dakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
       'Utah', 'Vermont', 'Virginia', 'Washington', 'West Virginia',
       'Wisconsin', 'Wyoming', 'District of Columbia', 'kentucky', nan,
       'Puerto Rico'], dtype=object)

## Sorting data

DataFrames/Series can be sorted using sort_values() method. Let us sort our dataset using total_inmate_cases in descending order.

In [11]:
prisonData.sort_values(by='total_inmate_cases',ascending=False)

nyt_id                                      facility_name  \
2247  E5DA4853                                     U.S. Marshalls   
1458  046ADCE7                                 Fresno County jail   
92    41367DB9                                Avenal State Prison   
120   19DE7187  Substance Abuse Treatment Facility and State P...   
106   550543A1                                     Soledad prison   
...        ...                                                ...   
2374  0A1D2EB9                                 Carver County Jail   
1924  C75A68C0                      Vance County Detention Center   
2372  3867C0A8                            Sheridan federal prison   
344   5E4670FA  Melbourne Center for Personal Growth youth center   
693   64DD6F6E   Flowood Community Work Center/Restitution Center   

                 facility_type facility_city facility_county  \
2247            U.S. Marshalls           NaN        Muskogee   
1458                      Jail        Fresno          Fresno   
92                State prison        Avenal           Kings   
120               State prison      Corcoran           Kings   
106               State prison       Soledad        Monterey   
...                        ...           ...             ...   
2374          Detention center        Chaska          Carver   
1924                      Jail     Henderson           Vance   
2372          Detention center      Sheridan         Yamhill   
344   State juvenile detention     Melbourne         Brevard   
693               State prison       Flowood           Hinds   

      facility_county_fips  facility_state  facility_lng  facility_lat  \
2247                 40101             NaN    -95.374390     35.749979   
1458                  6019      California   -119.789751     36.738211   
92                    6031      California   -120.123310     35.974332   
120                   6031      California   -119.554854     36.058048   
106                   6053      California   -121.382640     36.470807   
...                    ...             ...           ...           ...   
2374                 27019       Minnesota    -93.592635     44.787408   
1924                 37181  North Carolina    -78.407240     36.330646   
2372                 41071      Washington   -123.381623     45.084014   
344                  12009         Florida    -80.713169     28.122853   
693                  28049     Mississippi    -90.193159     32.296677   

      latest_inmate_population  max_inmate_population_2020  \
2247                       NaN                         NaN   
1458                    2042.0                         NaN   
92                      3469.0                      4158.0   
120                     4488.0                      4531.0   
106                     4374.0                         NaN   
...                        ...                         ...   
2374                       NaN                         NaN   
1924                       NaN                         NaN   
2372                       NaN                         NaN   
344                       12.0                         NaN   
693                        NaN                         NaN   

      total_inmate_cases  total_inmate_deaths  total_officer_cases  \
2247               12290                   31                    0   
1458                3814                    0                  171   
92                  3108                    8                  532   
120                 3011                    7                  636   
106                 2719                   18                  355   
...                  ...                  ...                  ...   
2374                   0                    0                    0   
1924                   0                    0                    3   
2372                   0                    0                    0   
344                    0                    0                    6   
693   

As you can see the U.S. Marshalls facility has the highest number of total_inmate_cases followed by Fresno County jail.

Can you sort the dataset using total_inmate_deaths in descending order. 

## Writing data to a file

We can use the to_csv() method to write a DataFrame to a CSV file. Let's write a subset of our file containing all the facilities with 1,000 or more inmate cases to a file called facilities_1000ormore.csv to the data folder.

In [12]:
#let's create the subset
prisonData1000OrMore = prisonData[prisonData['total_inmate_cases']>=1000]
prisonData1000OrMore

nyt_id                           facility_name   facility_type  \
34    7120943C         Goose Creek Correctional Center    State prison   
50    AA807759  Arizona State Prison Complex – Douglas    State prison   
51    91B22ADD                    Eyman Complex prison    State prison   
52    0CF96DB1     Arizona State Prison Complex – Yuma    State prison   
58    E830C278    Arizona State Prison Complex – Lewis    State prison   
...        ...                                     ...             ...   
2568  233CC887          Pollock federal prison complex  Federal prison   
2588  79BEB932               Seagoville federal prison  Federal prison   
2599  4A388A9A      Terre Haute federal prison complex  Federal prison   
2608  D9A37098           Tucson federal prison complex  Federal prison   
2609  20C2E350      Victorville federal prison complex  Federal prison   

     facility_city    facility_county  facility_county_fips facility_state  \
34         Wasilla  Matanuska-Susitna                  2170         Alaska   
50         Douglas            Cochise                  4003        Arizona   
51        Florence              Pinal                  4021        Arizona   
52        San Luis               Yuma                  4027        Arizona   
58         Buckeye           Maricopa                  4013        Arizona   
...            ...                ...                   ...            ...   
2568       Pollock              Grant                 22043      Louisiana   
2588    Seagoville             Dallas                 48113          Texas   
2599   Terre Haute               Vigo                 18167        Indiana   
2608        Tucson               Pima                  4019        Arizona   
2609   Victorville     San Bernardino                  6071     California   

      facility_lng  facility_lat  latest_inmate_population  \
34     -149.990022     61.360095                    1293.0   
50     -109.600147     31.445500                    1848.0   
51     -111.337569     33.031474                    5362.0   
52     -114.643082     32.494131                    4316.0   
58     -112.644365     33.201281                    4617.0   
...            ...           ...                       ...   
2568    -92.432853     31.465159                    2590.0   
2588    -96.567986     32.655933                    1798.0   
2599    -87.455019     39.410206                    2291.0   
2608   -110.865276     32.084730                    1586.0   
2609   -117.364873     34.567075                    3310.0   

      max_inmate_population_2020  total_inmate_cases  total_inmate_deaths  \
34                        1348.0                1041                    0   
50                        2132.0                1163                    0   
51                        5471.0                2023                    5   
52                        4862.0                2010                    6   
58                        4617.0                1310                    2   
...                          ...                 ...                  ...   
2568                      2754.0                1211                    0   
2588                      2004.0                1336                    5   
2599                      2624.0                1248                    6   
2608                      1931.0                1067                   10   
2609                      4210.0                1696                    3   

      total_officer_cases  total_officer_deaths note  
34                      0                   0.0  NaN  
50                      0                   0.0  NaN  
51                      0                   0.0  NaN  
52                      0                   0.0  NaN  
58                      0                   0.0  NaN  
...                   ...                   ...  ...  
2568                    0                   0.0  NaN  
2588                    0                   0.0  NaN  
2599             

In [13]:
prisonData1000OrMore.to_csv(r'data/facilities_1000ormore.csv',index=False)

In the next chapter we will look into advanced Pandas concepts such as Grouping and Joining. 